File: mobius-server/app/models/automation.py
import uuid
from datetime import datetime
from sqlalchemy import String, Boolean, DateTime, Integer, Text
from sqlalchemy.orm import Mapped, mapped_column
from app.models.base import Base


class Automation(Base):
    __tablename__ = "automations"
    id: Mapped[str] = mapped_column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    user_id: Mapped[str] = mapped_column(String, index=True)
    name: Mapped[str] = mapped_column(String, nullable=False)
    description: Mapped[str] = mapped_column(Text, default="")
    trigger_type: Mapped[str] = mapped_column(String, default="cron")  # "cron" | "manual" | "webhook"
    trigger_config: Mapped[str] = mapped_column(Text, default="{}")  # JSON: {"cron": "0 9 * * *"}
    script: Mapped[str] = mapped_column(Text, default="")  # Python source code
    status: Mapped[str] = mapped_column(String, default="draft")  # "draft" | "active" | "paused" | "error"
    last_run: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    last_result: Mapped[str | None] = mapped_column(Text, nullable=True)
    last_error: Mapped[str | None] = mapped_column(Text, nullable=True)
    run_count: Mapped[int] = mapped_column(Integer, default=0)
    created_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


# Backward-compat alias
ScheduledAutomation = Automation

File: mobius-app/lib/shared/models/automation.dart
class Automation {
  final String id;
  final String name;
  final String description;
  final String triggerType;
  final Map<String, dynamic> triggerConfig;
  final String script;
  final String status;
  final DateTime? lastRun;
  final String? lastResult;
  final String? lastError;
  final int runCount;

  const Automation({
    required this.id,
    required this.name,
    required this.description,
    required this.triggerType,
    required this.triggerConfig,
    required this.script,
    required this.status,
    this.lastRun,
    this.lastResult,
    this.lastError,
    this.runCount = 0,
  });

  /// Convenience getter for cron expression from triggerConfig.
  String get cronExpr {
    final cron = triggerConfig['cron'];
    return cron is String ? cron : '';
  }

  /// Backward-compatible active flag.
  bool get active => status == 'active';

  factory Automation.fromJson(Map<String, dynamic> json) => Automation(
        id: json['id'] as String,
        name: (json['name'] ?? '') as String,
        description: (json['description'] ?? '') as String,
        triggerType: (json['trigger_type'] ?? 'cron') as String,
        triggerConfig: json['trigger_config'] is Map
            ? Map<String, dynamic>.from(json['trigger_config'] as Map)
            : <String, dynamic>{},
        script: (json['script'] ?? '') as String,
        status: (json['status'] ?? 'active') as String,
        lastRun: json['last_run'] != null
            ? DateTime.parse(json['last_run'] as String)
            : null,
        lastResult: json['last_result'] as String?,
        lastError: json['last_error'] as String?,
        runCount: (json['run_count'] ?? 0) as int,
      );

  Map<String, dynamic> toJson() => {
        'id': id,
        'name': name,
        'description': description,
        'trigger_type': triggerType,
        'trigger_config': triggerConfig,
        'script': script,
        'status': status,
        'last_run': lastRun?.toIso8601String(),
        'last_result': lastResult,
        'last_error': lastError,
        'run_count': runCount,
      };
}
